File: scripts/integration_guidance_native.mjs
/*
 * Copyright (c) 2026 OceanBase.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Execute registered adapters against a controlled transport, without contacting a Server.
import { createInterface } from 'node:readline'
import { OPERATIONS } from '../integrations/dsh/plugins/powercontext/src/operations.generated.ts'

const lines = createInterface({ input: process.stdin, terminal: false })[Symbol.asyncIterator]()
const send = value => process.stdout.write(JSON.stringify(value) + '\n')
const receive = async () => {
  const line = await lines.next()
  if (line.done) throw new Error('Evaluation transport closed before replying')
  return JSON.parse(line.value)
}
const host = process.argv[2]
const scope = 'fixture-scope'
const signal = new AbortController().signal
const request = async (operation, payload) => {
  send({ kind: 'request', operation, payload })
  const reply = await receive()
  if (reply.kind !== 'response') throw new Error('Expected a controlled HTTP response')
  return { kind: 'json', status: OPERATIONS[operation].successStatuses[0], requestId: undefined, value: reply.value }
}
let execute
if (host === 'dsh') {
  const { registerTools } = await import('../integrations/dsh/plugins/powercontext/src/tools.ts')
  const tools = []
  registerTools({ tools: { register: tool => tools.push(tool) }, on: () => {} }, {
    client: { request }, resolveScope: async () => scope, config: {}, log: () => {},
  }, tool => tool)
  execute = (name, args) => {
    const tool = tools.find(tool => tool.name === name)
    if (!tool) throw new Error(`DSH has no tool ${name}`)
    return tool.execute(args, { signal })
  }
} else if (host === 'pi') {
  const { registerTools } = await import('../integrations/pi/plugins/powercontext/src/tools.ts')
  const tools = []
  registerTools({ registerTool: tool => tools.push(tool) }, {
    client: { request }, resolveScope: async () => scope, config: { maxBytes: 8000 },
  })
  execute = async (name, args) => {
    const tool = tools.find(tool => tool.name === name)
    if (!tool) throw new Error(`Pi has no tool ${name}`)
    const result = await tool.execute('evaluation', args, signal, () => {}, {
      cwd: process.cwd(), hasUI: true, ui: { confirm: async () => true },
    })
    return JSON.parse(result.content[0].text)
  }
} else if (host === 'opencode') {
  delete process.env.POWERCONTEXT_OPENCODE_ACTIVATION_PROBE_PATH
  delete process.env.POWERCONTEXT_OPENCODE_ACTIVATION_PROBE_NONCE
  process.env.POWERCONTEXT_OPENCODE_SCOPE_ID = scope
  process.env.POWERCONTEXT_OPENCODE_BASE_URL = 'http://127.0.0.1:1'
  const { OPERATIONS } = await import('../integrations/opencode/plugins/powercontext/src/operations.generated.ts')
  globalThis.fetch = async (url, init) => {
    const path = new URL(url).pathname
    if (path.endsWith('/scope-bindings/resolve')) return Response.json({ scope_id: scope })
    const operation = Object.entries(OPERATIONS).find(([, spec]) => spec.path === path && spec.method === init.method)?.[0]
    if (!operation) throw new Error(`Unexpected evaluation HTTP request: ${init.method} ${path}`)
    const result = await request(operation, JSON.parse(init.body))
    return Response.json(result.value, { status: result.status })
  }
  const { PowerContextPlugin } = await import('../integrations/opencode/plugins/powercontext/src/index.ts')
  const hooks = await PowerContextPlugin({ directory: process.cwd(), client: {
    session: { get: async () => ({ data: { directory: process.cwd() } }) }, app: { log: async () => ({}) },
  } })
  execute = async (name, args) => {
    if (!hooks.tool[name]) throw new Error(`OpenCode has no tool ${name}`)
    return JSON.parse(await hooks.tool[name].execute(args, {
      sessionID: 'evaluation', abort: signal, ask: async () => {},
    }))
  }
} else if (host === 'openclaw') {
  globalThis.fetch = async (url, init) => {
    const path = new URL(url).pathname
    if (path.endsWith('/scope-bindings/resolve')) return Response.json({ scope_id: scope })
    const operation = Object.entries(OPERATIONS).find(([, spec]) => spec.path === path && spec.method === init.method)?.[0]
    if (!operation) throw new Error(`Unexpected evaluation HTTP request: ${init.method} ${path}`)
    const result = await request(operation, JSON.parse(init.body))
    return Response.json(result.value, { status: result.status })
  }
  const { default: plugin } = await import('../integrations/openclaw/plugins/memory-powercontext/dist/index.js')
  const tools = []
  const context = { agentId: 'main', sessionKey: 'agent:main:telegram:direct:fixture' }
  plugin.register({
    config: {}, pluginConfig: { endpoint: 'http://127.0.0.1:1', scopeId: scope },
    runtime: { config: { current: () => ({}) }, agent: { session: { getSessionEntry: () => ({ chatType: 'direct' }) } } },
    registerTool: factory => { const tool = factory(context); if (tool) tools.push(tool) },
    registerMemoryCapability: () => {}, registerCommand: () => {}, registerService: () => {}, on: () => {},
    logger: { warn: () => {}, debug: () => {}, info: () => {} },
  })
  execute = async (name, args) => {
    const tool = tools.find(tool => tool.name === name)
    if (!tool) throw new Error(`OpenClaw has no tool ${name}`)
    return JSON.parse((await tool.execute('evaluation', args, signal)).content[0].text)
  }
} else {
  throw new Error(`Unsupported native adapter: ${host}`)
}
send({ kind: 'ready' })
while (true) {
  const line = await lines.next()
  if (line.done) break
  try {
    const { name, arguments: args } = JSON.parse(line.value)
    send({ kind: 'result', value: await execute(name, args) })
  } catch (error) {
    send({ kind: 'error', message: String(error) })
  }
}

File: scripts/integration_guidance_native.py
# Copyright (c) 2026 OceanBase.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Native adapter execution with controlled client replies; never contact a Server."""

from __future__ import annotations

import asyncio
import json
import os
import shutil
from pathlib import Path
from typing import Any

from integration_guidance_handoff import HandoffFixture

NATIVE_HOSTS = {"dsh", "pi", "opencode", "openclaw"}
ROOT = Path(__file__).resolve().parent.parent


class NativeHandoffSession:
    def __init__(self, host: str) -> None:
        self.host = host
        self.process: asyncio.subprocess.Process | None = None
        self.requests: list[dict[str, Any]] = []

    async def _receive(self) -> dict[str, Any]:
        if self.process is None or self.process.stdout is None:
            message = "Native adapter stdout is unavailable"
            raise RuntimeError(message)
        line = await asyncio.wait_for(self.process.stdout.readline(), timeout=60)
        if not line:
            message = f"{self.host} evaluation adapter exited before returning a result"
            raise RuntimeError(message)
        return json.loads(line)

    async def _send(self, value: dict[str, Any]) -> None:
        if self.process is None or self.process.stdin is None:
            message = "Native adapter stdin is unavailable"
            raise RuntimeError(message)
        self.process.stdin.write((json.dumps(value) + "\n").encode())
        await self.process.stdin.drain()

    async def call(self, name: str, arguments: dict[str, Any], fixture: HandoffFixture) -> dict[str, Any]:
        if self.process is None:
            node = os.environ.get("POWERCONTEXT_GUIDANCE_NODE") or shutil.which("node")
            if node is None:
                message = "Native guidance evaluation requires Node.js 22.19+ and installed host dependencies"
                raise RuntimeError(message)
            self.process = await asyncio.create_subprocess_exec(
                node,
                "--experimental-transform-types",
                str(Path(__file__).with_suffix(".mjs")),
                self.host,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                # Leave startup diagnostics visible; never mix them into model tool replies.
                cwd=ROOT,
            )
            if (await self._receive()).get("kind") != "ready":
                message = f"{self.host} evaluation adapter did not initialize"
                raise RuntimeError(message)
        await self._send({"name": name, "arguments": arguments})
        requests_before = len(self.requests)
        while True:
            message = await self._receive()
            if message["kind"] == "request":
                self.requests.append(message)
                result = fixture.respond(message["operation"], message["payload"])
                await self._send({"kind": "response", "value": result})
            elif message["kind"] == "result":
                failed = (
                    message["value"].get("unavailable") if self.host == "openclaw" else not message["value"].get("ok")
                )
                if len(self.requests) != requests_before + 1 or failed:
                    message = f"{self.host} {name} did not complete one controlled operation: {message['value']}"
                    raise ValueError(message)
                return message["value"]
            else:
                message = f"{self.host} {name}: {message.get('message', 'invalid adapter reply')}"
                raise RuntimeError(message)

    async def close(self) -> None:
        if self.process is not None:
            if self.process.returncode is None:
                self.process.kill()
            await self.process.wait()
